File: articles/biology-scientific-understanding-living-order/r/growth_models.R
# Growth models for living order in R.

growth_path <- file.path("data", "growth_observations.csv")
logistic_path <- file.path("data", "logistic_scenarios.csv")

if (!file.exists(growth_path)) {
  growth_path <- file.path("..", "data", "growth_observations.csv")
  logistic_path <- file.path("..", "data", "logistic_scenarios.csv")
}

growth <- read.csv(growth_path)

fit_condition <- function(df) {
  fit <- lm(log(abundance) ~ time, data = df)

  r_est <- coef(fit)[["time"]]
  n0_est <- exp(coef(fit)[["(Intercept)"]])
  td_est <- log(2) / r_est

  data.frame(
    condition = unique(df$condition),
    growth_rate = r_est,
    initial_abundance = n0_est,
    doubling_time = td_est,
    r_squared_log_space = summary(fit)$r.squared
  )
}

fit_df <- do.call(
  rbind,
  lapply(split(growth, growth$condition), fit_condition)
)

print(round(fit_df, 5))

logistic_growth <- function(time, initial_abundance, growth_rate, carrying_capacity) {
  carrying_capacity /
    (1 + ((carrying_capacity - initial_abundance) / initial_abundance) *
      exp(-growth_rate * time))
}

scenarios <- read.csv(logistic_path)

rows <- lapply(seq_len(nrow(scenarios)), function(i) {
  s <- scenarios[i, ]
  time <- seq(0, s$time_end, by = s$dt)

  trajectory <- logistic_growth(
    time,
    s$initial_abundance,
    s$growth_rate,
    s$carrying_capacity
  )

  data.frame(
    scenario = s$scenario,
    final_abundance = tail(trajectory, 1),
    fraction_of_capacity = tail(trajectory, 1) / s$carrying_capacity
  )
})

summary_df <- do.call(rbind, rows)

print(round(summary_df, 5))

File: articles/biology-scientific-understanding-living-order/python/growth_models.py
"""
Growth models for living order.

Run:
    python python/growth_models.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from living_order_core import fit_exponential_growth, logistic_growth


ARTICLE_DIR = Path(__file__).resolve().parents[1]
GROWTH_PATH = ARTICLE_DIR / "data" / "growth_observations.csv"
LOGISTIC_PATH = ARTICLE_DIR / "data" / "logistic_scenarios.csv"


def main() -> None:
    """Fit exponential growth and simulate logistic constraint scenarios."""

    growth = pd.read_csv(GROWTH_PATH)

    fit_rows = []

    for condition, group in growth.groupby("condition"):
        fit = fit_exponential_growth(
            group["time"].to_numpy(dtype=float),
            group["abundance"].to_numpy(dtype=float),
        )

        fit_rows.append(
            {
                "condition": condition,
                "growth_rate": fit.growth_rate,
                "initial_abundance": fit.initial_abundance,
                "doubling_time": fit.doubling_time,
                "r_squared_log_space": fit.r_squared_log_space,
            }
        )

    logistic_scenarios = pd.read_csv(LOGISTIC_PATH)

    summary_rows = []

    for _, scenario in logistic_scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        trajectory = logistic_growth(
            time=time,
            initial_abundance=scenario["initial_abundance"],
            growth_rate=scenario["growth_rate"],
            carrying_capacity=scenario["carrying_capacity"],
        )

        summary_rows.append(
            {
                "scenario": scenario["scenario"],
                "initial_abundance": scenario["initial_abundance"],
                "growth_rate": scenario["growth_rate"],
                "carrying_capacity": scenario["carrying_capacity"],
                "final_abundance": trajectory[-1],
                "fraction_of_capacity": trajectory[-1] / scenario["carrying_capacity"],
            }
        )

    print(pd.DataFrame(fit_rows).round(5).to_string(index=False))
    print(pd.DataFrame(summary_rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
